File: solution/structure/LRU.go
package structure

import "errors"

type node[V any] struct {
    val V
    next *node[V]
    prev *node[V]
}

type LRU[K comparable, V any] struct {
    forward_map map[K]*node[V]
    reverse_map map[*node[V]]K
    head *node[V]
    tail *node[V]
    Size int
    Capc int
}

func NewLRU[K comparable, V any](capc int) (*LRU[K, V], error) {
    if capc == 0 { return nil, errors.New("Invalid Capacity") }

    var e error

    if capc == 1 { e = errors.New("Warning: LRU's capacity is set to 1") }

    return &LRU[K, V]{
        make(map[K]*node[V], capc),
        make(map[*node[V]]K, capc),
        nil,
        nil,
        0,
        capc,
    }, e
}

func (l *LRU[K, V]) Get(k K) (V, bool) {
    var v V; if l.Size == 0 { return v, false }

    if node, in := l.forward_map[k]; in {
        if node == l.head { return node.val, true }

        if node == l.tail {
            l.tail = l.tail.prev; l.tail.next = nil
        } else {
            node.prev.next, node.next.prev = node.next, node.prev
        }

        node.next, l.head.prev, l.head = l.head, node, node;

        return node.val, true
    }

    return v, false
}

func (l *LRU[K, V]) Update(k K, v V) {
    if l.Size == 0 { l.prepend(k, v); return }

    if node, in := l.forward_map[k]; in {
        node.val = v

        if node == l.head { return }

        if node == l.tail {
            l.tail = l.tail.prev; l.tail = nil
        } else {
            node.prev.next, node.next.prev = node.next, node.prev
        }

        node.next, l.head.prev, l.head = l.head, node, node;

        return
    }

    l.prepend(k, v)
}

func (l *LRU[K, V]) prepend(k K, v V) {
    node := &node[V]{v, nil, nil}; 

    l.forward_map[k] = node; l.reverse_map[node] = k

    if l.Size == 0 { l.head, l.tail = node, node; l.Size++; return }

    if l.Size == l.Capc {
        tail_key := l.reverse_map[l.tail]; delete(l.forward_map, tail_key); 

        delete(l.reverse_map, l.tail)

        if l.Capc == 1 {
            l.tail.next, l.tail.prev = nil, nil; l.head, l.tail = nil, nil
        } else {
            l.tail = l.tail.prev; l.tail.next = nil;
        }
    } else {
        l.Size++
    }

    if l.Capc == 1 && l.Capc == l.Size { 
        l.head, l.tail = node, node
    } else {
        node.next, l.head.prev = l.head, node; l.head = node
    }
}

File: src/solutions/LRU.ts
type Node<V> = {
    v: V
    next: Node<V> | undefined; 
    prev: Node<V> | undefined;
}

export default class LRU<K, V> {
    private map: Map<K, Node<V>>;
    private reverse_map: Map<Node<V>, K>;
    private head: Node<V> | undefined;
    private tail: Node<V> | undefined;
    private capc: number;
    private size: number;

    public constructor(capc: number) {
        this.map = new Map<K, Node<V>>();
        this.reverse_map = new Map<Node<V>, K>();
        this.head = undefined;
        this.tail = undefined;
        this.capc = capc;
        this.size = 0;
    }

    public get(k: K): V | undefined {
        if (this.size === 0) return undefined;

        if (!this.map.has(k)) return undefined;

        const node = this.map.get(k) as Node<V>;

        if (node === this.head) return node.v;

        if (node === this.tail) {
            if (!node.prev) throw Error("Exception occurs");

            this.tail = node.prev; 
            this.tail.next = undefined;
        } else {
            if (!node.prev) throw Error("Exception occurs");
            if (!node.next) throw Error("Exception occurs");

            node.prev.next = node.next; 
            node.next.prev = node.prev;
        }

        if (!this.head) throw Error("Exception occurs");

        this.head.prev = node;
        node.next = this.head;
        this.head = node;

        return node.v;
    }

    public update(k: K, v: V): void {
        if (this.size === 0) { this.prepend(k, v); return; }

        if (!this.map.has(k)) { this.prepend(k, v); return; }

        const node = this.map.get(k) as Node<V>;
        node.v = v;

        if (node === this.head) return; 

        if (node === this.tail) {
            if (!node.prev) throw Error("Exception occurs");

            this.tail = node.prev;
            this.tail.next = undefined;
        } else {
            if (!node.prev) throw Error("Exception occurs");
            if (!node.next) throw Error("Exception occurs");

            node.prev.next = node.next;
            node.next.prev = node.prev;
        }

        if (!this.head) throw Error("Exception occurs");

        this.head.prev = node;
        node.next = this.head;
        this.head = node;
    }

    private prepend(k: K, v: V): void {
        const node: Node<V> = { v: v, next: undefined, prev: undefined };
        
        this.map.set(k, node);
        this.reverse_map.set(node, k);

        if (this.size === 0) { 
            this.head = this.tail = node;
            this.size = this.size + 1;
            return;
        }

        if (this.size === this.capc) {
            if (!this.tail) throw Error("Exception occurs");
            
            if (!this.reverse_map.has(this.tail)) throw ("Exception occurs");

            const tail_key = this.reverse_map.get(this.tail) as K;

            this.reverse_map.delete(this.tail);
            this.map.delete(tail_key);

            if (this.capc === 1) {
                this.head = this.tail = undefined;
            } else {
                if (!this.tail.prev) throw Error("Exception occurs");

                this.tail = this.tail.prev;
                this.tail.next = undefined;
            }
        } else {
            this.size = this.size + 1;
        }

        if (!this.head) throw Error("Exception occurs"); 

        node.next = this.head;
        this.head.prev = node;
        this.head = node; 
    }
}
